internal/cmd/devices/list.go

```diff
@@ -0,0 +1,272 @@
+	"github.com/saucelabs/saucectl/internal/usage"
+	"github.com/spf13/cobra"
+)
+
+const (
+	JSONOutput = "json"
+	TextOutput = "text"
+)
+
+type DeviceWithStatus struct {
+	ID     string `json:"id"`
+	Name   string `json:"name"`
+	OS     string `json:"os"`
+	Status string `json:"status"`
+}
+
+type Filter struct {
+	Name   string
+	Os     string
+	Status string
+}
+
+type ListOptions struct {
+	Page         int
+	Size         int
+	Status       bool
+	OutputFormat string
+	Filter       Filter
+}
+
+func ListCommand() *cobra.Command {
+	var out string
+	var page int
+	var size int
+	var nameFilter string
+	var osFilter string
+	var addStatus bool
+	var statusFilter string
+
+	cmd := &cobra.Command{
+		Use: "list",
+		Aliases: []string{
+			"ls",
+		},
+		Short:        "Returns the list of devices",
+		SilenceUsage: true,
+		PreRun: func(cmd *cobra.Command, _ []string) {
+			tracker := usage.DefaultClient
+
+			go func() {
+				tracker.Collect(
+					cmds.FullName(cmd),
+					usage.Flags(cmd.Flags()),
+				)
+				_ = tracker.Close()
+			}()
+		},
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			if page < 0 {
+				return errors.New("invalid page")
+			}
+			if size < 0 {
+				return errors.New("invalid size")
+			}
+			if out != JSONOutput && out != TextOutput {
+				return errors.New("unknown output format")
+			}
+
+			if statusFilter != "" {
+				addStatus = true
+			}
+
+			options := ListOptions{
+				Page:         page,
+				Size:         size,
+				Status:       addStatus,
+				OutputFormat: out,
+				Filter: Filter{
+					Name:   nameFilter,
+					Os:     osFilter,
+					Status: statusFilter,
+				},
+			}
+
+			return list(cmd.Context(), options)
+		},
+	}
+
+	flags := cmd.PersistentFlags()
+	flags.StringVarP(&out, "out", "o", "text", "OutputFormat format to the console. Options: text, json.")
+	flags.IntVarP(&page, "page", "p", 0, "Page for pagination. Default is 0.")
+	flags.IntVarP(&size, "size", "s", 0, "Per page for pagination. Default is all.")
+	flags.StringVarP(&nameFilter, "name", "n", "", "Filter devices by name.")
+	flags.StringVar(&osFilter, "os", "", "Filter devices by OS.")
+	flags.BoolVar(&addStatus, "statuses", false, "Fetch status for devices.")
+	flags.StringVar(&statusFilter, "status", "", "Filter devices by status. Implies --statuses if not set.")
+
+	return cmd
+}
+
+func list(ctx context.Context, options ListOptions) error {
+	devs, err := devicesReader.GetDevices(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get devices: %w", err)
+	}
+
+	var devsWithStatuses []DeviceWithStatus
+	if options.Status {
+		res, err := getDevicesWithStatuses(ctx, devs)
+		if err != nil {
+			return fmt.Errorf("failed to get devices: %w", err)
+		}
+		devsWithStatuses = res
+	} else {
+		devsWithStatuses = getDevicesWithEmptyStatuses(devs)
+	}
+
+	var filtered = filterDevices(devsWithStatuses, options.Filter)
+
+	var paginated []DeviceWithStatus
+	var from, to int
+	if options.Size > 0 {
```

Comment:
If the API doesn't require this is it even worth exposing page / sizes client side? 🤔

Reply:
I guess not... removed.